ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -216,6 +216,9 @@ private void waitForCommit() throws InterruptedException {
     // Thus it is possible to have applied > committed initially.
     final long applied = getLastAppliedIndex();
     for(; applied >= raftLog.getLastCommittedIndex() && state == State.RUNNING && !shouldStop(); ) {
+      if (server.getSnapshotRequestHandler().getPending().get().isPresent()) {
```

Comment:
We shall use `server.getSnapshotRequestHandler().shouldTriggerTakingSnapshot()` in this scenario, see  https://github.com/apache/ratis/blob/17c965292133114d6169feb19f32670e2b122032/ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotManagementRequestHandler.java#L112-L114 This method will clear the flag and guarantee one snapshot be taken each request. Otherwise, one request may trigger two snapshots.

Reply:
@SzyWilliam , it is a bug in my suggestion but not in @OneSizeFitsQuorum 's early change.  In the early change, it calls `shouldTakeSnapshot()` and then `server.getSnapshotRequestHandler().shouldTriggerTakingSnapshot()`.